src/AutoRest.CSharp/LowLevel/Output/OperationMethodChainBuilder.cs

```diff
@@ -94,12 +102,41 @@ public LowLevelClientMethod BuildOperationMethodChain()
             return new LowLevelClientMethod(protocolMethodSignature, convenienceMethod, _restClientMethod, requestBodyType, responseBodyType, diagnostic, _protocolMethodPaging, Operation.LongRunning, _conditionHeaderFlag);
         }
 
-        private bool ShouldConvenienceMethodGenerated(ReturnTypeChain returnTypeChain)
+        private bool ShouldGenerateConvenienceMethod()
         {
             return Operation.GenerateConvenienceMethod
                 && (!Operation.GenerateProtocolMethod
-                || _orderedParameters.Where(parameter => parameter.Convenience != KnownParameters.CancellationTokenParameter).Any(parameter => !IsParameterTypeSame(parameter.Convenience, parameter.Protocol))
-                || !returnTypeChain.Convenience.Equals(returnTypeChain.Protocol));
+                ||_orderedParameters.Where(parameter => parameter.Convenience != KnownParameters.CancellationTokenParameter).Any(parameter => !IsParameterTypeSame(parameter.Convenience, parameter.Protocol))
+                || !_returnType.Convenience.Equals(_returnType.Protocol));
+        }
+
+        private bool HasAmbiguityBetweenProtocolAndConvenience()
+        {
+            return _orderedParameters.Where(parameter => parameter.Convenience != KnownParameters.CancellationTokenParameter).All(parameter => IsParameterTypeHasValueOverlap(parameter.Convenience, parameter.Protocol));
+        }
+
+        private bool ShouldRequestContextOptional()
+        {
+            if (!ShouldGenerateConvenienceMethod())
+            {
+                return true;
+            }
+
+            if (_sourceInputModel != null)
+            {
+                var existingMethod = _sourceInputModel.FindMethod(_namespaceName, _clientName, Operation.CleanName, _orderedParameters.Select(p => p.Protocol).WhereNotNull().Select(p => p.Type));
+                if (existingMethod != null && existingMethod.Parameters.Length > 0 && existingMethod.Parameters.Last().IsOptional)
+                {
+                    return true;
+                }
+            }
+
+            if (HasAmbiguityBetweenProtocolAndConvenience() && Configuration.AmbiguityHandling == Configuration.AmbiguityHandlingOption.UseOverload)
+            {
+                return false;
+            }
+
+            return true;
         }
 
         private bool IsParameterTypeSame(Parameter? first, Parameter? second)
```

Comment:
is this method correct?
when first is null and second is NOT null, we return true, meaning that they have the same type?
This does not look correct to me.